### ISSUE TITLE
Expected behavior and bugs in editing failed proofs (edit_at request in XML protocol)

### ISSUE THREAD

**Issue by jfehrle** [coq/coq#15961]
As a first step, I need to understand how the feature is supposed to work for various cases.  Then figure out what's been broken in 8.15.

Attn: @gares

**Comment by jfehrle** [comment]
Case 1: 8.13.2:  If I add a space as shown, is the second line supposed to be retracted (no longer green)?

![image](https://user-images.githubusercontent.com/1253341/165567964-9606027f-9259-4f0d-a2df-316a3c150492.png)

![image](https://user-images.githubusercontent.com/1253341/165568136-d27edc1e-708c-4910-ba4b-c07605bc4e0d.png)

**Comment by jfehrle** [comment]
Case 2A: 8.13.2: Selecting "exact I." and pressing CTRL-d (Tools/Comment) omits the closing " *)".  Clearly a bug.

![image](https://user-images.githubusercontent.com/1253341/165574238-de49c56a-32e0-49f8-a755-f3b447a28049.png)

![image](https://user-images.githubusercontent.com/1253341/165574816-000a2816-02ff-4a8c-a65b-569615c8beef.png)

Case 2B: If I then press CTRL-Z (undo in the editor), the buffer is corrupted:

**Comment by jfehrle** [comment]
Case 3: 8.13.2: Add a space in the comment, then delete the second line gives "coqidetop died badly" dialog

![image](https://user-images.githubusercontent.com/1253341/165584427-664c8238-3977-4bfb-9ed4-71aa25f0e44a.png)

**Comment by jfehrle** [comment]
@ppedrot Do you recall what you were doing when you got "coqidetop died" or saw warnings highlighted in the wrong place?  Were you editing a failed proof or unprocessed text?

**Comment by jfehrle** [comment]
Case 5: in master.  Missing tooltips for warnings.  IIRC there should be a tooltip on `minus_Sn_m`.  (probably not related to edit_at.)

![image](https://user-images.githubusercontent.com/1253341/165598504-a5a8148f-3134-40ec-bc9d-eeb6676732fb.png)

**Comment by gares** [comment]
I think you shoul play with proof which are at least 2 steps, even idtac is fine. Otherwise you are in weird corner cases, which may be buggy, but for sure don't affect users that much

**Comment by jfehrle** [comment]
> you need to start the block with a Proof. statement for it to be well-bracketed.

I'd expect `Proof.` would be highlighted here.  When Coq finishes processing and a successful proof, should the highlight on `Qed` be removed?  Once Coq is finished, if you edit a successful proof, should it invalidate everything after it?

![image](https://user-images.githubusercontent.com/1253341/165845060-b37c9267-074b-4dbf-8aba-08d89a1e857a.png)

**Comment by jfehrle** [comment]
The asynchronous execution (showing "foo" before "done") is by design, right?  If so, another thing that should be described in the doc since it's helpful for users.

![image](https://user-images.githubusercontent.com/1253341/165846343-be76b0d6-b2a3-4256-a414-0d21ed6344d5.png)

**Comment by Alizter** [comment]
> The asynchronous execution (showing "foo" before "done") is by design, right? If so, another thing that should be described in the doc since it's helpful for users.
> 
This might not always be the case however. Since the main process steps down, it is able to print foo first. The proof of C is checked by a proof worker process. In large proofs, it could be the case that the proof worker finishes first.

**Comment by Alizter** [comment]
> @gares Are you supposed to be able to edit in multiple locations? If you do, then Coq is supposed to process the first one before the second one?
> 
> Case 6 master: Inconsistent behavior: result depends on order of editing failed sections
> 

The first behaviour is the expected one. The second seems to be a bug.

**Comment by jfehrle** [comment]
> Since the main process steps down, it is able to print foo first.

I think both were in proof workers.  There were 2 lemmas before line 3.

**Comment by Alizter** [comment]
@jfehrle Since there is no Qed at the end of the document, I don't think the last one spawned a proof worker.

**Comment by jfehrle** [comment]
Case 7: While waiting for "Run to End" to complete, I added a space on line 2, which immediately aborted the running proof (and removed highlights after that point).  No so user friendly.  I thought we wanted to disallow changes in the proved part while Coq is still running?

![image](https://user-images.githubusercontent.com/1253341/165849249-0cb1d7ec-1243-4371-80da-7ececa9ee88f.png)

**Comment by jfehrle** [comment]
> Since there is no Qed at the end of the document, I don't think the last one spawned a proof worker.

Same result with Qed.  I updated the screen shot.

**Comment by gares** [comment]
having the qed fail because the proof is incomplete is a corner case, maybe buggy, but I think you shold use as the exampl:
```
Lemma x: True /\ True.
Proof.
split.
exact 2. (* error *)
exact I.
Qed.
```

**Comment by gares** [comment]
In particular the comment you put before qed is part of the qed sentence itself, this is a bit special IIRC.

**Comment by jfehrle** [comment]
Case 8: Handling of commands within async proofs (corner case)

Commands in the middle of proofs are executed immediately, in order, before async processing proceeds?  It's odd to see `Ltac ...` marked as unprocessed when it clearly has been.  This is by design?  This could be mentioned in the doc.  The highlighting could be adjusted, too.  I expect that's not difficult.  OTOH, I expect users rarely put commands in the middle of proofs.

![image](https://user-images.githubusercontent.com/1253341/166116991-969fa77b-c865-4b87-8c3a-2e24310cda2c.png)

**Comment by gares** [comment]
Yes, commands are extruded from proofs, or better duplicate and run on both the worker and master.
I believe the desired semantics is that they are only run by the worker and their effects don't escape the proof. I would not color them in green.
